### PR TITLE
Issue 2223: Fix StateSynchronizer compact should invoke truncate

### LINKED ISSUE #2223: StateSynchronizer compact should invoke truncate
**Problem description**
Once data has been compacted there is little point in retaining it. 

**Problem location**
StateSynchronizerImpl

**Suggestions for an improvement**
Call RevisionedStreamClient.truncate

### PR DESCRIPTION
**Pull request by huide9**
Signed-off-by: huide9 <huide9@gmail.com>

**Change log description**  
At the end of compact operation, after compareAndSetMark(), delete the old data through the revision (oldMark here).

**Purpose of the change**  
Fix #2223.

**What the code does**  
After data has been compacted, delete old data through the given revision.

reference from the documentation:
http://pravega.io/docs/latest/state-synchronizer/#delete-operations
```
You can think of compact() as a form of "garbage collection" in StateSynchronizer.  After a 
certain number of changes have been written to SharedState, it might be efficient to write 
out a new initial state, an accumulated representation of all the changes, to the Stream.  
That way data older than the compact operation can be ignored and eventually removed 
from the Stream.
```

**How to verify it**  
All the existing tests should continue to pass like before.

### CODE REVIEW

**Comment by huide9** [comment]
> We need a new unit test to verify this behavior.
> In particular we need a test that has two "hosts" where one writes several updates and another reads some but not all of these. Then the first host compacts the updates. The second "host" should be able to fetch updates and arrive at the new state.

testCompactWithTrunction() added for such case.

**Comment by huide9** [comment]
[Issue 2223 StateSynchronizer compact should invoke truncate](https://trello.com/c/uOi71SjW/156-issue-2223-statesynchronizer-compact-should-invoke-truncate)

File: client/src/test/java/io/pravega/client/state/impl/SynchronizerTest.java [hunk 1]
```diff
@@ -411,5 +411,108 @@ public void testSetOperations() throws EndOfSegmentException {
...
+                new JavaSerializer<>(),
+                SynchronizerConfig.builder().build());
+
+        assertEquals(0, syncA.bytesWrittenSinceCompaction());
+        assertEquals(0, syncB.bytesWrittenSinceCompaction());
+
+        AtomicInteger callCount = new AtomicInteger(0);
+
+        syncA.initialize(new RegularUpdate("a"));
+        syncB.initialize(new RegularUpdate("a"));
```
[the remark this answers is not in the document]
Comment (huide9) [reply client/src/test/java/io/pravega/client/state/impl/SynchronizerTest.java]: I guess you're talking syncB.initialize(new RegularUpdate("a")); 
                            =>  syncB.initialize(new RegularUpdate("b"));          
`initialize` will init it with a given value if there isn't, or it do nothing. I actually use `A` before, but changed it back to b later.

File: client/src/test/java/io/pravega/client/state/impl/SynchronizerTest.java [hunk 2]
```diff
@@ -411,5 +411,108 @@ public void testSetOperations() throws EndOfSegmentException {
...
+        syncA.updateState((state, updates) -> {
+            callCount.incrementAndGet();
+            updates.add(new RegularUpdate("c"));
+        });
+        assertEquals("c", syncA.getState().getValue());
+        syncB.fetchUpdates();
+        assertEquals("c", syncB.getState().getValue());
+        assertTrue(syncA.bytesWrittenSinceCompaction() > size);
+
+        syncA.updateState((state, updates) -> {
```
[the remark this answers is not in the document]
Comment (huide9) [reply client/src/test/java/io/pravega/client/state/impl/SynchronizerTest.java]: yes, the purpose is to have more updates and every time syncB can get the updated value.

File: client/src/test/java/io/pravega/client/state/impl/SynchronizerTest.java [hunk 3]
```diff
@@ -411,5 +411,108 @@ public void testSetOperations() throws EndOfSegmentException {
...
+            callCount.incrementAndGet();
+            updates.add(new RegularUpdate("d"));
+        });
+        assertEquals("d", syncA.getState().getValue());
+        syncB.fetchUpdates();
+        assertEquals("d", syncB.getState().getValue());
+        assertTrue(syncA.bytesWrittenSinceCompaction() > size);
+
+        syncA.updateState((state, updates) -> {
```
[the remark this answers is not in the document]
Comment (huide9) [reply client/src/test/java/io/pravega/client/state/impl/SynchronizerTest.java]: let's keep one dup step

it still can match to:
` where one writes several updates and another reads some but not all of these.`

File: client/src/test/java/io/pravega/client/state/impl/SynchronizerTest.java [hunk 4]
```diff
@@ -411,5 +411,108 @@ public void testSetOperations() throws EndOfSegmentException {
...
+
+        AtomicInteger callCount = new AtomicInteger(0);
+
+        syncA.initialize(new RegularUpdate("a"));
+        syncB.initialize(new RegularUpdate("a"));
+        assertEquals("a", syncA.getState().getValue());
+        assertEquals("a", syncB.getState().getValue());
+
+        syncA.updateState((state, updates) -> {
+            callCount.incrementAndGet();
```
[the remark this answers is not in the document]
Comment (huide9) [reply client/src/test/java/io/pravega/client/state/impl/SynchronizerTest.java]: if you mean:   assertEquals(2, callCount.get());  it fixed in new commit

File: client/src/test/java/io/pravega/client/state/impl/SynchronizerTest.java [hunk 5]
```diff
@@ -411,5 +411,121 @@ public void testSetOperations() throws EndOfSegmentException {
...
+            updates.add(new RegularUpdate("f"));
+        });
+        assertEquals(5, callCount.get());
+        assertEquals("f", syncA.getState().getValue());
+        assertTrue(syncA.bytesWrittenSinceCompaction() > size);
+
+        assertEquals("e", syncB.getState().getValue());
+        syncB.compact(state -> {
+            callCount.incrementAndGet();
+            return new RegularUpdate("e");
```
Comment (tkaitchuck) [inline client/src/test/java/io/pravega/client/state/impl/SynchronizerTest.java]: This seems wrong. It's writing "e" to compact a state that is currently "f". Instead do `new RegularUpdate(state.getValue())`.

**Pull request merged**